src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DecisionDagRewriter.cs

```diff
@@ -410,53 +412,510 @@ private void GenerateTest(BoundExpression test, BoundDecisionDagNode whenTrue, B
+                        }
+                    case BoundDagValueTest value:
+                        {
+                            // Gather up the (value, label) pairs, starting with the first one
+                            loweredNodes.Add(testNode);
+                            var cases = ArrayBuilder<(ConstantValue value, LabelSymbol label)>.GetInstance();
+                            cases.Add((value: value.Value, label: GetDagNodeLabel(testNode.WhenTrue)));
+                            BoundTestDecisionDagNode previous = testNode;
+                            while (previous.WhenFalse is BoundTestDecisionDagNode p &&
+                                p.Test is BoundDagValueTest vd &&
+                                vd.Input.Equals(input) &&
+                                !this._dagNodeLabels.ContainsKey(p) &&
+                                !loweredNodes.Contains(p))
+                            {
+                                cases.Add((value: vd.Value, label: GetDagNodeLabel(p.WhenTrue)));
+                                loweredNodes.Add(p);
+                                previous = p;
+                            }
+
+                            var otherwise = GatherValueDispatchNodes(previous.WhenFalse, loweredNodes, input, fac);
+                            return PushEqualityTestsIntoTree(value.Syntax, otherwise, cases.ToImmutableAndFree(), fac);
+                        }
+                    default:
+                        {
+                            var label = GetDagNodeLabel(node);
+                            return new ValueDispatchNode.LeafDispatchNode(node.Syntax, label);
+                        }
                 }
+            }
+
+            /// <summary>
+            /// Push the set of equality tests down to the level of the leaves in the value dispatch tree.
+            /// </summary>
+            private ValueDispatchNode PushEqualityTestsIntoTree(
+                SyntaxNode syntax,
+                ValueDispatchNode otherwise,
+                ImmutableArray<(ConstantValue value, LabelSymbol label)> cases,
+                IValueSetFactory fac)
+            {
+                if (cases.IsEmpty)
+                    return otherwise;
 
-                // Gather up the (value, label) pairs, starting with the first one
-                var cases = ArrayBuilder<(ConstantValue value, LabelSymbol label)>.GetInstance();
-                cases.Add((value: firstTest.Value, label: GetDagNodeLabel(firstTestNode.WhenTrue)));
+                switch (otherwise)
+                {
+                    case ValueDispatchNode.LeafDispatchNode leaf:
+                        return new ValueDispatchNode.SwitchDispatch(syntax, cases, leaf.Label);
+                    case ValueDispatchNode.SwitchDispatch sd:
+                        return new ValueDispatchNode.SwitchDispatch(sd.Syntax, sd.Cases.Concat(cases), sd.Otherwise);
```

Comment:
>sd.Syntax [](start = 68, length = 9)

`syntax`?

Reply:
No.  The idea is to preserve the error location, for example, if `bool operator==(string, string)` is found to be missing.  The choice to use `sd.Syntax` instead of `syntax` is based on the behavior of tests, some of which would otherwise report the missing member on the whole switch expression's syntax rather than on just the constant pattern.

---
In reply to: [391109113](https://github.com/dotnet/roslyn/pull/42313#discussion_r391109113) [](ancestors = 391109113)

Reply:
In my opinion no error should be reported on the entire switch syntax if we're doing that for some other diagnostics, that doesn't seem to be helpful when it's highlighted in IDE. perhaps just the `switch` token?

Reply:
I do not believe there are any scenarios where do report an error on the entire switch syntax.